### PR TITLE
Accept react nodes as children

### PR DESCRIPTION
**Pull request by Elbarae1921**
This pull request fixes the issue #73 

Currently the component does not take into account children. I've changed it so that it strips the children of any non-string elements and replaces them with a placeholder, then after the string output is emojified, it puts the elements back in place.

I've tested this with [linkify-react](https://www.npmjs.com/package/react-linkify) and it works.

I'm not sure if this is the right way to do it or not (because it just ignores anything that isn't plain text) so I'm waiting for your feedback.

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -1,3 +1,46 @@
...
+
+export function stripNonStringElements(children) {
+  if (Array.isArray(children)) {
+    const stripped = children
+      .map(x => (typeof x === "string" ? x : NON_STRING_PLACEHOLDER))
+      .join("");
+    return [stripped, children.filter(x => typeof x !== "string")];
+  } else if (typeof children === "string") {
+    return [children, []];
+  }
```
Comment (enzoferey) [inline src/utils.js:15]: I fear we miss out on some values here being too restrictive about only 2 possibilities for the `children` (and I believe React `children` could be other types than `Array` or `string`).

I believe we should be doing:

```js
if (Array.isArray(children)) {
    ....
}

return [children, []];
```

File: src/utils.js [hunk 2]
```diff
@@ -1,3 +1,46 @@
...
+}
+
+export function returnStrippedElements(stripped, elements) {
+  if (!stripped || !Array.isArray(stripped)) return stripped;
```
Comment (enzoferey) [inline src/utils.js:19]: Same issue here I believe, `stripped` might be of other types than `Array` or `string` ?

File: src/utils.js [hunk 3]
```diff
@@ -1,3 +1,46 @@
...
+  if (!elements || !elements.length) return stripped;
+
+  let count = -1;
+  return stripped.flatMap(x => {
```
Comment (enzoferey) [inline src/utils.js:23]: I'm not sure if using `flatMap` is fine regarding browser support, I would probably change it with a `reduce` + `map` alternative. Do you mind ?

Also this function is a bit convoluted with up to 6 levels of nesting, can we do better for future contributors / maintainers ? 😁

**Review by enzoferey — COMMENTED** [review]
Hey @Elbarae1921 !

Thanks for your PR 🙏🏻  It looks good, I have left a few comments you can find here below.

Regarding your concerns about this being fine or not, I believe we should just add a few more tests and see what happens. Looking at `React.ReactNode` type:

```ts
type ReactNode = ReactElement | string | number | ReactFragment | ReactPortal | boolean | null | undefined;
```

Let's add a test that adds all these as `children` and see what we get out 👍🏻

File: src/utils.js (hunk 1 above)
Comment (Elbarae1921) [reply src/utils.js:15]: I did some more tests and I think `children` is always an array, the contents of the array can either be a string or an object, sometimes a null. The only case where `children` is not an array is when it's undefined (e.g no children were passed).
So I believe we should remove the condition outside of the `stripNonStringElements` function and instead check if children is an array before calling the function. If `children` is not an array then `text` should stay as it is.
Maybe throw an error if both `text` and `children` are undefined.

File: src/utils.js (hunk 3 above)
Comment (Elbarae1921) [reply src/utils.js:23]: You're right, I'm used to the typescript compiler taking care of that for me. (maybe there's some babel config that can be tuned).
I'll see about splitting the function as well.

File: src/utils.js (hunk 1 above)
Comment (enzoferey) [reply src/utils.js:15]: All right, I think best thing we can do is write test cases for each of those cases and see how it behaves 😁

**Comment by Elbarae1921** [comment]
Hello, been a while 😃. I've changed some things, did more testing (manually) and it seems to be working. Now it throws an error if neither the `text` nor any children were passed, I've also changed the type definition to match that logic. I tried to split up the `returnStrippedElements` function as it was getting a bit too big. And finally I've replaced the native `Array.prorotype.flatMap` method with a homemade `flatMap` function.

I still ought to write some more tests for those edge cases you mentioned [here](https://github.com/tommoor/react-emoji-render/pull/92#pullrequestreview-958474344) tho.

**Comment by enzoferey** [comment]
Hey @Elbarae1921 ! I appreciate the effort 🙏🏻 

Let's see those automated tests indeed so we can reach a good solution through them 👍🏻

**Comment by Elbarae1921** [comment]
Hello @enzoferey, I've added some tests. Could you please take a look and see if those are enough or if you have some more in mind that we should add.

**Comment by enzoferey** [comment]
Thanks @Elbarae1921 🙌🏻 

I have approved the Github Actions run and the tests pass, so that's good. I will check the code when I have some time in the upcoming days to get this ready to merge 🚀

File: index.d.ts [hunk 4]
```diff
@@ -22,27 +23,31 @@ declare module 'react-emoji-render' {
     options?: Options;
   }
 
+  type RequireProperty<T, Prop extends keyof T> = T & {[key in Prop]-?:T[key]};
+
+  type PropsRequireTextOrChildren = RequireProperty<Props, 'text'> | RequireProperty<Props, 'children'>;
```
Comment (enzoferey) [inline index.d.ts:28]: Pretty clean 👌🏻 

The types were updated at https://github.com/tommoor/react-emoji-render/pull/99, so you will need to take into account these changes, but this approach is perfect.

File: src/__tests__/index-test.js [hunk 5]
```diff
@@ -16,6 +17,58 @@ import renderer from "react-test-renderer";
       expect(tree).toMatchSnapshot();
     });
 
+    test("emoji as children", () => {
```
Comment (enzoferey) [inline src/__tests__/index-test.js:20]: The tests are very on point 👌🏻 

I wonder what would happen if we put some emoji inside a nested HTML element on inside a nested React component. We should probably add a test case for each of these.

File: src/renderer.js [hunk 6]
```diff
@@ -108,8 +110,21 @@ export default function Emoji({
   onlyEmojiClassName,
   options = {},
   className,
+  children,
   ...rest
 }) {
+  let non_string_elements = [];
```
Comment (enzoferey) [inline src/renderer.js:116]: Please use camelCase naming => `nonStringElements`.

File: src/renderer.js [hunk 7]
```diff
@@ -108,8 +110,21 @@ export default function Emoji({
...
+
+  if (!!children && Array.isArray(children)) {
+    const [strippedChildren, elements] = stripNonStringElements(children);
+    text = strippedChildren;
+    non_string_elements = elements;
+  } else if (!!children && typeof children === "string") {
+    text = children;
+  } else if (!text) {
+    throw new Error("Emoji: either children or text prop must be provided");
```
Comment (enzoferey) [inline src/renderer.js:125]: How about using the library name as a prefix ? So it's easier to spot:

```suggestion
    throw new Error("react-emoji-render: either children or text prop must be provided");
```

File: src/renderer.js [hunk 8]
```diff
@@ -120,7 +135,11 @@ export default function Emoji({
     return true;
   }
 
-  const output = toArray(text, options);
+  let output = returnStrippedElements(
```
Comment (enzoferey) [inline src/renderer.js:138]: We don't need to switch to `let` here as we don't reassign.

Comment (enzoferey) [inline src/renderer.js:138]: How about naming the function `returnNonStringStrippedElements` ? I would say that's a more accurate description of the function's logic and it matches better the opposite of the `stripNonStringElements` function.

File: src/utils.js [hunk 9]
```diff
@@ -1,3 +1,65 @@
...
+
 export function escapeStringToBeUsedInRegExp(string) {
   return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"); // $& means the whole matched string
 }
+
+export function stripNonStringElements(children) {
+  const stripped = children
+    .map(x => (typeof x === "string" ? x : NON_STRING_PLACEHOLDER))
+    .join("");
+  return [stripped, children.filter(x => typeof x !== "string")];
```
Comment (enzoferey) [inline src/utils.js:11]: Let's be a bit more explicit on the variables names.

```suggestion
  const strippedChildren = children
    .map(child => (typeof child === "string" ? child : NON_STRING_PLACEHOLDER))
    .join("");
  const elements = children.filter(x => typeof x !== "string");
  return [strippedChildren, elements];
```

File: src/utils.js [hunk 10]
```diff
@@ -1,3 +1,65 @@
...
+        else {
+          count++;
+          return [elements[count], y];
+        }
+      });
+    } else {
+      count++;
+      return [...split, elements[count]];
+    }
+  }
```
Comment (enzoferey) [inline src/utils.js:52]: Same here, a bit more explicitness:

```suggestion
  function replacePlaceholder(string) {
    const split = string.split(NON_STRING_PLACEHOLDER);
    if (split.length > 1) {
      return flatMap(split, (item, index) => {
        if (index === 0) return [item];
        else {
          count++;
          return [elements[count], item];
        }
      });
    } else {
      count++;
      return [...split, elements[count]];
    }
  }
```

File: src/utils.js [hunk 11]
```diff
@@ -1,3 +1,65 @@
...
+    if (typeof x === "string") {
+      if (x.trim() === NON_STRING_PLACEHOLDER) {
+        count++;
+        return elements[count];
+      } else if (x.includes(NON_STRING_PLACEHOLDER)) {
+        return replacePlaceholder(x);
+      }
+    }
+    return x;
+  });
```
Comment (enzoferey) [inline src/utils.js:64]: And finally:

```suggestion
  return flatMap(stripped, item => {
    if (typeof item === "string") {
      if (item.trim() === NON_STRING_PLACEHOLDER) {
        count++;
        return elements[count];
      } else if (item.includes(NON_STRING_PLACEHOLDER)) {
        return replacePlaceholder(item);
      }
    }
    return item;
  });
```

**Review by enzoferey — COMMENTED** [review]
Thanks again @Elbarae1921 for the work. Please check my comments below to close this up 👍🏻

File: README.md [hunk 12]
```diff
@@ -45,6 +45,9 @@ import { Twemoji } from 'react-emoji-render';
 
 <Twemoji text="This ❤️ sentence includes :+1: a variety of emoji types :)" />
 
+// or with children
+<Twemoji>This ❤️ sentence includes :+1: a variety of emoji types :)</Twemoji>
```
[the remark this answers is not in the document]
Comment (Elbarae1921) [reply README.md:49]: agreed.

File: src/__tests__/index-test.js (hunk 5 above)
Comment (Elbarae1921) [reply src/__tests__/index-test.js:20]: Right now it only recognizes emojis that are direct children (as every non-string element is simply ignored). I think it would be quite complicated to add this functionality for nested emojis.

**Comment by Elbarae1921** [comment]
@enzoferey I've added the changes you've requested, could you please take another look? There's a good chance I messed up something as I've struggled a bit with the rebase lol.

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (enzoferey) [reply README.md:54]: Sorry if I was not clear about it, I meant adding it to the "Basic Usage" section AND removing it from the `Twemoji` and `Emojione` components docs, that way we avoid repeating the different options over and over again.

File: src/__tests__/index-test.js (hunk 5 above)
Comment (enzoferey) [reply src/__tests__/index-test.js:20]: Understood. I would still add some test cases to reflect this design decision, do you mind ?

**Comment by enzoferey** [comment]
Amazing @Elbarae1921 ! Indeed I think you missed a few little details, let me link them to you as bullet points 🙌🏻 

- https://github.com/tommoor/react-emoji-render/pull/92#discussion_r950924387
- https://github.com/tommoor/react-emoji-render/pull/92#discussion_r950925042
- https://github.com/tommoor/react-emoji-render/pull/92#discussion_r950758323

Thank you again for your effort, let's merge this as soon as those are solved 🚀

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (Elbarae1921) [reply README.md]: yeah sorry, that's a result of my botched rebase lol. fixed now.

File: src/__tests__/index-test.js (hunk 5 above)
Comment (Elbarae1921) [reply src/__tests__/index-test.js:20]: Done.

**Comment by enzoferey** [comment]
Great @Elbarae1921 💯 Did you happen to forget about https://github.com/tommoor/react-emoji-render/pull/92#discussion_r950758323 ? Last one 🙌🏻

**Comment by Elbarae1921** [comment]
> Great @Elbarae1921 💯 Did you happen to forget about [#92 (comment)](https://github.com/tommoor/react-emoji-render/pull/92#discussion_r950758323) ? Last one 🙌🏻

@enzoferey I did make those changes, the snippet there is outdated.

**Comment by enzoferey** [comment]
Ah right, I got confused because `x` argument was not renamed to `child` as in the suggestion, so it was half done 🙈  Do you mind renaming it ?

**Comment by Elbarae1921** [comment]
@enzoferey my bad, didn't notice the `x`. Should be good now.

**Review by enzoferey — APPROVED** [review]

**Comment by enzoferey** [comment]
You are the man @Elbarae1921 😁 

Thank you for the effort. Merging 🚀 

cc @tommoor any chances we could get a new package version published ?
